microsoft/testsuites/performance/networkperf.py

```diff
@@ -194,17 +340,228 @@ def perf_tcp_pps(
         )
-        if isinstance(client.capability.network_interface.data_path, NetworkDataPath):
-            data_path = client.capability.network_interface.data_path.value
-        information: Dict[str, str] = environment.get_information()
-        pps_message = dict_to_fields(information, pps_message)
-        pps_message.test_execution_tag = test_case_name
-        pps_message.data_path = data_path
         notifier.notify(pps_message)
+
+    def perf_ntttcp(
+        self,
+        log: Logger,
+        environment: Environment,
+        udp_mode: bool = False,
+        connections: Optional[List[int]] = None,
+    ) -> None:
+        client = cast(RemoteNode, environment.nodes[0])
+        server = cast(RemoteNode, environment.nodes[1])
+        test_case_name = inspect.stack()[1][3]
+        perf_tuning: Dict[str, List[Dict[str, str]]] = {
+            client.name: [],
+            server.name: [],
+        }
+        sys_list = [
+            {"kernel.pid_max": "122880"},
+            {"vm.max_map_count": "655300"},
+            {"net.ipv4.ip_local_port_range": "1024 65535"},
+        ]
+        if connections is None:
+            connections = NTTTCP_TCP_CONCURRENCY
+            if udp_mode:
+                sys_list = [
+                    {"net.core.rmem_max": "67108864"},
+                    {"net.core.rmem_default": "67108864"},
+                    {"net.core.wmem_default": "67108864"},
+                    {"net.core.wmem_max": "67108864"},
+                ]
+                connections = NTTTCP_UDP_CONCURRENCY
+        try:
+            client_ntttcp = client.tools[Ntttcp]
+            server_ntttcp = server.tools[Ntttcp]
+            client_lagscope = client.tools[Lagscope]
+            server_lagscope = server.tools[Lagscope]
+            stop_firewall(environment)
+            set_systemd_tasks_max([client, server], log)
+            for node in [client, server]:
+                sysctl = node.tools[Sysctl]
+                for tcp_sys in sys_list:
+                    for variable, value in tcp_sys.items():
+                        perf_tuning[node.name].append({variable: sysctl.get(variable)})
+                        sysctl.write(variable, value)
+            data_path = get_nic_datapath(client)
+            server_nic_name = server.nics.default_nic
+            client_nic_name = client.nics.default_nic
+            dev_differentiator = "Hypervisor callback interrupts"
+            if NetworkDataPath.Sriov.value == data_path:
+                server_nic_name = server.nics.get_lower_nics()[0]
+                client_nic_name = client.nics.get_lower_nics()[0]
+                dev_differentiator = "mlx"
+            server_lagscope.run_as_server(ip=server.internal_address)
+            max_server_threads = 64
+            perf_ntttcp_message_list: List[Any] = []
+            for test_thread in connections:
+                if test_thread < max_server_threads:
```

Comment:
Add comments to explain the relationship of thread and buffer size.

Reply:
I am not sure the relationship.

Reply:
Nobody knows more than you. The logic set num_threads and buffer_size together. Please try to explain as possible as you can in next PR.